source/log.c

```diff
@@ -0,0 +1,386 @@
+/*
+* Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Licensed under the Apache License, Version 2.0 (the "License").
+* You may not use this file except in compliance with the License.
+* A copy of the License is located at
+*
+*  http://aws.amazon.com/apache2.0
+*
+* or in the "license" file accompanying this file. This file is distributed
+* on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+* express or implied. See the License for the specific language governing
+* permissions and limitations under the License.
+*/
+
+#ifndef _CRT_SECURE_NO_WARNINGS
+	#define _CRT_SECURE_NO_WARNINGS
+#endif
+
+#ifndef _CRT_NONSTDC_NO_DEPRECATE
+	#define _CRT_NONSTDC_NO_DEPRECATE
+#endif
+
+#include <aws/common/log.h>
+
+#include <aws/common/atomic.h>
+#include <aws/common/common.h>
+#include <aws/common/memory_pool.h>
+#include <aws/common/mutex.h>
+#include <aws/common/thread.h>
+
+
+#include <stdio.h>
+#include <time.h>
+#include <inttypes.h>
+
+#define AWS_SINGLY_LIST_REVERSE(T, list) \
+    do { \
+        T* reverse_node__ = 0; \
+        while (list) { \
+            T* reverse_next__ = list->next; \
+            list->next = reverse_node__; \
+            reverse_node__ = list; \
+            list = reverse_next__; \
+        } \
+        list = reverse_node__; \
+    } while (0)
+
+#define AWS_ATOMIC_SINGLY_LIST_INSERT(head, node) \
+    do { \
+        do { \
+            (node)->next = (head); \
+        } while (!aws_atomic_compare_exchange_ptr((void **)&(head), (void **)&((node)->next), (node))); \
+    } while (0)
+
+struct aws_log_message {
+    struct aws_log_message *next;
+};
+
+struct aws_log_context {
+    struct aws_log_message *message_list;
+    struct aws_log_message *delete_list;
+    size_t max_message_len;
+    int table_index;
+    struct aws_memory_pool* message_pool;
+    struct aws_allocator *alloc;
+};
+
+enum aws_log_entry_state {
+    AWS_LOG_ENTRY_STATE_NO_WRITERS,
+    AWS_LOG_ENTRY_STATE_WRITER,
+    AWS_LOG_ENTRY_STATE_DELETEME,
+};
+
+struct aws_log_context_entry {
+    int state;
+    struct aws_log_context *ctx;
+};
+
+struct aws_log_system_params {
+    size_t max_message_len;
+    int memory_pool_message_count;
+    int max_message_count;
+    struct aws_allocator *alloc;
+};
+
+static void s_aws_log_default_report_function(const char *log_message) {
+    (void)log_message;
+}
+
+static int s_log_message_count;
+static int s_log_system_running;
+static struct aws_log_system_params s_log_system_params;
+
+static int s_log_context_count;
+static struct aws_log_context_entry s_log_table[AWS_LOG_MAX_LOGGING_THREADS];

```

Comment:
Fixed size was for simplicity of the system. Originally it was a doubly linked list behind a lock. The lock turned out to be a bit nasty when performing clean up operations on threads, and I switched to this table idea instead. I think 256 threads is a perfectly acceptable limit that nobody will hit. Though, it could be changed to a dynamically allocated resizeable buffer (with a mutex), if we so wanted. Though I do think this would be a massive over-engineering.

Reply:
When I say "not particularly unreasonable", I mean that applications may well have that many threads in some cases.